lib/go/thrift/binary_protocol.go

```diff
@@ -479,38 +486,21 @@ func (p *TBinaryProtocol) readAll(ctx context.Context, buf []byte) (err error) {
 	return NewTProtocolException(err)
 }
 
-const readLimit = 32768
-
 func (p *TBinaryProtocol) readStringBody(size int32) (value string, err error) {
-	if size < 0 {
-		return "", nil
-	}
-
-	var (
-		buf bytes.Buffer
-		e   error
-		b   []byte
-	)
+	buf, err := safeReadBytes(size, p.trans)
+	return string(buf), NewTProtocolException(err)
+}
 
-	switch {
-	case int(size) <= len(p.buffer):
-		b = p.buffer[:size] // avoids allocation for small reads
-	case int(size) < readLimit:
-		b = make([]byte, size)
-	default:
-		b = make([]byte, readLimit)
+// This function is shared between TBinaryProtocol and TCompactProtocol.
+//
+// It tries to read size bytes from trans, in a way that prevents large
+// allocations when size is insanely large (mostly caused by malformed message).
```

Comment:
How does it actually prevent large allocations? Previously there was readLimit

Reply:
It's in the details of the implementation of `io.CopyN` and `bytes.Buffer`. what they did is similar to the previous version, that they will do the copy chunk by chunk (and grow the buffer when necessary), instead of allocate the whole size up front.

Reply:
if `size` is large you end up with a huge allocated array no matter _how you copy data into it_. The main issue in the ticket is that the value of `size` could be corrupted and be greater than the total message size.

Reply:
No. With `io.CopyN` + `bytes.Buffer` it stops when there's no more data to read. See this line in the benchmark result I attached to the commit message:

```
BenchmarkSafeReadBytes/max-size-12                462532              2494 ns/op           14464 B/op          7 allocs/op
```

The size in that call is max int32, but it only allocated 14KB of memory.

Reply:
Got it. Thanks!